File: gto-support-api-base/src/main/java/org/ccci/gto/android/common/api/Session.java
package org.ccci.gto.android.common.api;

import android.content.SharedPreferences;
import android.text.TextUtils;

import java.util.Arrays;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Object representing an individual session for this API. Can be extended to track additional session data.
 */
public class Session {
    protected static final String PREF_SESSION_BASE_NAME = "session";

    @NonNull
    private final String mBaseAttrName;

    @Nullable
    public final String id;

    public Session(@Nullable final String id) {
        this(id, PREF_SESSION_BASE_NAME);
    }

    public Session(@Nullable final String id, @Nullable final String baseAttrName) {
        mBaseAttrName = baseAttrName != null ? baseAttrName : PREF_SESSION_BASE_NAME;
        this.id = id;
    }

    public Session(@NonNull final SharedPreferences prefs) {
        this(prefs, null);
    }

    public Session(@NonNull final SharedPreferences prefs, @Nullable final String baseAttrName) {
        mBaseAttrName = baseAttrName != null ? baseAttrName : PREF_SESSION_BASE_NAME;
        this.id = prefs.getString(getPrefAttrName("id"), null);
    }

    public boolean isValid() {
        return this.id != null;
    }

    @NonNull
    public final String getPrefAttrName(@NonNull final String type) {
        return mBaseAttrName + "." + type;
    }

    public void save(@NonNull final SharedPreferences.Editor prefs) {
        prefs.putString(getPrefAttrName("id"), this.id);
    }

    public void delete(@NonNull final SharedPreferences.Editor prefs) {
        prefs.remove(getPrefAttrName("id"));
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final Session that = (Session) o;
        return TextUtils.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(new Object[] {id});
    }
}

File: gto-support-api-base/src/main/kotlin/org/ccci/gto/android/common/api/Session.kt
package org.ccci.gto.android.common.api

import android.content.SharedPreferences

private const val PREF_ID = "id"

/**
 * Object representing an individual session for this API. Can be extended to track additional session data.
 */
open class Session protected constructor(
    prefs: SharedPreferences? = null,
    id: String? = null,
    private val baseAttrName: String = PREF_SESSION_BASE_NAME
) {
    companion object {
        const val PREF_SESSION_BASE_NAME = "session"
    }

    @JvmField
    val id = id ?: prefs?.getString(prefAttrName(PREF_ID), null)

    @JvmOverloads
    constructor(
        id: String?,
        baseAttrName: String = PREF_SESSION_BASE_NAME
    ) : this(prefs = null, id = id, baseAttrName = baseAttrName)

    @JvmOverloads
    constructor(
        prefs: SharedPreferences,
        baseAttrName: String = PREF_SESSION_BASE_NAME
    ) : this(prefs = prefs, id = null, baseAttrName = baseAttrName)

    protected fun prefAttrName(type: String) = "$baseAttrName.$type"

    open fun save(prefs: SharedPreferences.Editor) {
        prefs.putString(prefAttrName(PREF_ID), id)
    }

    open fun delete(prefs: SharedPreferences.Editor) {
        prefs.remove(prefAttrName(PREF_ID))
    }

    open val isValid get() = id != null

    override fun equals(other: Any?) = when {
        this === other -> true
        !(other is Session && javaClass == other.javaClass) -> false
        else -> id == other.id
    }

    override fun hashCode() = id?.hashCode() ?: 0
}
